### ISSUE TITLE
RangeError: Maximum call stack size exceeded when making cy.request which times out

### ISSUE THREAD

**Issue by t00** [cypress-io/cypress#15629]
### Current behavior

When making cy.request(url) to the endpoint which does not respond in 30 seconds (default response timeout), cypress crashes with ```RangeError: Maximum call stack size exceeded``` error in socket.io-parser, used by bluebird library.

### Desired behavior
cy.request(url) should fail the test instead of crashing the cypress.

### Test code to reproduce

spec.js:

```js
/// <reference types="cypress" />

context('Network Requests', () => {
  it('cy.request() - make an XHR request', () => {
    // https://on.cypress.io/request
    cy.request('http://localhost:8080').then(r => {
      cy.wrap(r.body).should('contain', 'Hello');
    })
  })
})
```

server.js:

```js
const express = require('express');
const app = express();

app.get('/', async (req, res) => {
  await sleep(50000);
  res.send('Hello from App Engine!');
});

// Listen to the App Engine-specified port, or 8080 otherwise
const PORT = process.env.PORT || 8080;
app.listen(PORT, () => {
  console.log(`Server listening on port ${PORT}...`);
});

function sleep(ms) {
    return new Promise((resolve) => {
      setTimeout(resolve, ms);
    });
  } 
```

### Versions

Definitely 6.6.0 and 6.8.0, possibly earlier.

### Callstack

```
RangeError: Maximum call stack size exceeded
    at _deconstructPacket (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:21:28)
    at _deconstructPacket (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:40:32)
    at _deconstructPacket (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:40:32)
    at _deconstructPacket (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:40:32)
    at _deconstructPacket (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:40:32)

... 15k lines omitted ...

    at _deconstructPacket (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:40:32)
    at _deconstructPacket (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:40:32)
    at _deconstructPacket (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:40:32)
    at _deconstructPacket (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:40:32)
    at _deconstructPacket (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:32:26)
    at Object.deconstructPacket (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\binary.js:16:17)
    at Encoder.encodeAsBinary (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\index.js:81:41)
    at Encoder.encode (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io-parser\dist\index.js:43:29)
    at Client._packet (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io\dist\client.js:167:44)
    at Socket.packet (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io\dist\socket.js:161:21)
    at C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\socket\node_modules\socket.io\dist\socket.js:270:18
    at C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\server\lib\socket-base.js:304:28
    at tryCatcher (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\util.js:16:23)
    at Promise._settlePromiseFromHandler (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:547:31)
    at Promise._settlePromise (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:604:18)
    at Promise._settlePromise0 (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:649:10)
    at Promise._settlePromises (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\promise.js:725:18)
    at _drainQueueStep (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\async.js:93:12)
    at _drainQueue (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\async.js:86:9)
    at Async._drainQueues (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\async.js:102:5)
    at Immediate.Async.drainQueues [as _onImmediate] (C:\Users\Tests\AppData\Local\Cypress\Cache\6.6.0\Cypress\resources\app\packages\server\node_modules\bluebird\js\release\async.js:15:14)
    at processImmediate (internal/timers.js:456:21)
```

Edit: Changed ```r.body.should('contain', 'Hello');``` to ```cy.wrap(r.body).should('contain', 'Hello');```

**Comment by t00** [comment]
I uploaded a whole test project to: https://github.com/t00/cypress-stack-size-exceeded
to reproduce:

```
npm i
npm start &
npm run cypress:open
```

**Comment by jennifer-shehane** [comment]
Getting the following error running the code as provided. 

<img width="572" alt="Screen Shot 2021-03-24 at 3 15 17 PM" src="https://user-images.githubusercontent.com/1271364/112370498-db3a5580-8cb3-11eb-9a49-246543c4b0b9.png">

**Comment by t00** [comment]
Sorry, when testing I did not even get to this point - please change the line to ```cy.wrap(r.body).should('contain', 'Hello')```.
The test should not reach this point anyway as the request should time out after 30 seconds and server.js returns this response after 50 seconds.

I also updated github repo with the fix. Does it mean you have a working fix? That would be brilliant.

Also, when running server it is better to open a new terminal and run ```npm start``` in it separately instead of running in background using ```npm open```. I don't code for node personally.

**Comment by t00** [comment]
@jennifer-shehane Just to confirm - this test should fail - request timed out. Currently on cypress 6.8.0 it crashes cypress after 30 seconds.

**Comment by creage** [comment]
@t00 could it platform specific? You might want to add info about env specs you face with this issue on.

**Comment by t00** [comment]
@creage unlikely it is platform specific although I am not sure how cypress handles ajax response timeout on different platforms so this would make a difference. If there are problems to reproduce it, from my experience adding explicit ```"responseTimeout": 30000``` to cypress.json should force it to wait only 30 seconds.

I am able to reproduce the problem on our pipeline Windows 10 test machines and my personal Arch Linux desktop dev machine.

**Comment by t00** [comment]
@creage I might add we are using cypress installer of Chromium 88 for Windows machines and cypress 6.8.0 provided Electron on Linux - this should not make any difference as the crash happens in node.
Windows machine is running node 14.12, Linux is running node 15.12.

**Comment by JasonFairchild** [comment]
I believe I'm having this issue, in completely different scenario. I get to the point where the cy.request times out and the test fails, but some amount of time later, Cypress always crashes. For apparently no reason, but now that I found this issue I think it is the same. The only consistent part is that the request times out some time close to the crash. I can even start rerunning the test, but it still crashes several seconds later. This is really really annoying since I have a site I'm testing that often times out on the first time I run the test (probably for multiple reasons, but mostly because our test server resources are terrible and services may not be very optimized). Anyway, since I have to npx cypress open every time after it crashes, I am stuck only being able to run the test for the first time, so it crashes every time. 

Other odd thing is this test used to never crash. Perhaps it could be related to 6.8, but pretty sure I had this test not crash on that version as well. Dev just made changes to the test server though, so really a lot of variables at play. I will try running the same test on some other envs and then maybe try reverting back to cypress 6.7.x

Side note: For w/e reasons, Cypress seems to be much faster/more performant when running tests after the first launch of the browser and more. I've always wondered why? Especially since Cypress supposedly does all sorts of clearing before every test run, no matter how it is initiated. Some type of caching that doesn't get cleared (not browser cache)? More perhaps? But it is very consistent behavior, even though it is not exact. Meaning I've noticed tests running faster in 'open' mode after you have run the test all the way through at least once. Consistent across many types of differences including tech stacks/frameworks, test types, browser version and more.

**Comment by t00** [comment]
@JasonFairchild It has been confirmed in another thread that reverting to cypress@6.4.0 should fix the crashing issue. I personally simply increased a timeout as we are not testing response time for similar reasons to yours - test machines are not powerful enough. I am hoping cypress team is on the case (I reported it through support).
For your first start long response issue, you can also use some warming up of your api - either using i.e. npm wait-on, curl or other method - all before starting tests.

**Comment by t00** [comment]
@JasonFairchild also cypress can wait for response  with custom timeout:
```

context('Cypress.config()', () => {
  it('Get and set configuration options', () => {
    // this will change the config for the rest of your tests!
    //Cypress.config('responseTimeout', 40000);
    //expect(Cypress.config('responseTimeout')).to.eq(40000);

    cy.visit('https://independer.free.beeceptor.com/my/api/path').as('slowy'); // API Path is setup to respond after 35seconds
    cy.wait('@slow', { responseTimeout: 40000 });
  })
})
```

**Comment by JasonFairchild** [comment]
Ah, ok. Thanks for the quick heads up. Yeah, changing the response timeout seems simple and wouldn't hurt in my case since performance is purposely being left by the wayside at this time on this project (kind of sad, but that is another topic =). Honestly, I'm going to try simply aborting the test before it actually times out, since I'm watching anyway, and see if that makes any difference, haha. I suspect I could revert back to 6.4 also, but have to double check the changelog. I hope it gets fixed or improved soon as well. Seems like one of those things that could be happening to a lot more people and they don't understand why, or could become worse in a hurry, and become a real nuisance/flaky.

**Comment by bahmutov** [comment]
@t00 what OS shows the crash? I am trying to run the tests locally on Mac, without crashes.

**Comment by t00** [comment]
@bahmutov For me both Linux Arch and Windows 10. Crash will only happen when request timeout is enforced (default settings, 30 seconds).

**Comment by t00** [comment]
@bahmutov Try adding `Cypress.config('responseTimeout', 30000);` at the beginning of the test and/or add `"responseTimeout": 30000` to cypress.json

**Comment by t00** [comment]
@bahmutov Let me know if this works, would be interesting to know that on Mac default responseTimeout setting of 30 seconds is not enforced which might help if we ever decide to perform automated tests on Macs.

**Comment by bahmutov** [comment]
I don't think the response timeout is not enforced on Mac - I mean the test fails with the "no response within the timeout" error

**Comment by JasonFairchild** [comment]
For me, I would always have to wait a little while, like 10-15 seconds after Cypress reported they request timeout, before the crash would happen. Maybe it is longer for machines with more memory or something, or maybe it doesn't reproduce at all with enough memory? That last would seem odd to me though.

**Comment by t00** [comment]
I gave a go trying to reproduce the problem on a bare debian install and surprisingly it did not crash.
I tried node 12 (`curl -fsSL https://deb.nodesource.com/setup_12.x | bash -`) and latest lts (14).
It did not crash on Firefox, Electron nor Chromium (`apt install chromium chromium-l10n`).

If anyone has some ideas please share.

1. Get https://cdimage.debian.org/debian-cd/current/amd64/iso-cd/debian-10.9.0-amd64-netinst.iso
2. Create a new VM (on Linux I used virt-manager) - 4GB RAM, 4 CPU, 16GB HDD
3. Boot ISO, choose English, UK (maybe not relevant)
4. All default settings, use whole disk, confirm writing changes to disk, choose closest mirror
5. Select Xfce (small DE) while keeping other default selections
6. Select /dev/vda to install grub on a whole disk
7. Reboot, sign in
8. Open terminal and run following commands:

```
su
apt install -y git build-essential curl
curl -fsSL https://deb.nodesource.com/setup_lts.x | bash -
apt install -y nodejs
exit
git clone https://github.com/t00/cypress-stack-size-exceeded.git
cd cypress-stack-size-exceeded
npm i
npm start &
npm run cypress:open
```

Select Electron 78 and run test.

**Comment by t00** [comment]
@bahmutov @JasonFairchild GOT IT - all aforementioned browsers are affected.
After starting the test, within controlled browser window stop the test and start about 10 times (I tried every 1 second - total 10 seconds for restarts), leave running and cypress crashes after about 20 seconds - even before 30 second timeout!
Clearly some connection state clearing is not working and in real tests some previous test connections are left stale during a new spec.
Hope it helps.

**Comment by bahmutov** [comment]
Tried Docker `npx cypress run` mode, lots of tests, no crash. I think it might be headed interactive mode only?

**Comment by t00** [comment]
Originally the problem happened for me during standard integration cypress run tests, it apparently happens to be just easier to reproduce in interactive mode. Are you able to reproduce it on Mac using 10 times start/stop in interactive mode?
Looks like the issue is about multiple open connections stacking up during a single browser session and likely happens when one of them is being closed by a remote end while rest are waiting for a response.
Not sure about numbers, but certainly aforementioned 10 artificial connections are enough to get it broken.

**Comment by t00** [comment]
My suspicion is also about not properly disposing subscription to a connection somewhere in cypress internals, which when left hanging and eventually failing, tries to do something unexpected.

**Comment by JasonFairchild** [comment]
That sounds like it lines up with what I was doing. When the request would time out, I would rerun the test, anyway. But to be clear, my devs have since increased caching and performance of the app all around, so I basically never see request timeouts under the same circumstances any longer. Everything I'm saying is from back when I added my first comment.

**Comment by DVGY** [comment]
@t00 I have the same problem.

**Comment by t00** [comment]
@DVGY Looks like @bahmutov is from the cypress team and hopefully might be able to help :smile:

**Comment by bahmutov** [comment]
If only I could recreate this problem on Mac or via Docker container running locally and so far no luck

**Comment by t00** [comment]
Have you tried this (or variation - more stop/start, faster) on Mac?

```
git clone https://github.com/t00/cypress-stack-size-exceeded.git
cd cypress-stack-size-exceeded
npm i
npm start &
npm run cypress:open
```

_After starting the test, within controlled browser window stop the test and start about 10 times (I tried every 1 second - total 10 seconds for restarts), leave running and cypress crashes after about 20 seconds - even before 30 second timeout!_

**Comment by jennifer-shehane** [comment]
Duplicate of https://github.com/cypress-io/cypress/issues/15101 which we have a reproducible example for now. 

@t00 I could also get yours to reproduce after a few tries of stop/starting. 

Closing as Duplicate to track in one issue.
